### PR TITLE
Deploy-monitoring + alerts

### PR DESCRIPTION
**Pull request by adonagy**

### CODE REVIEW

File: apps/deploy_monitoring/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,33 @@
+[package]
+name = "deploy-monitoring"
+version = "1.2.0"
+authors = ["Adrian Nagy <adrian.nagy@simplestaking.com>"]
```
Comment (bkontur) [inline apps/deploy_monitoring/Cargo.toml:4]: @adonagy 
new email? :D

File: apps/deploy_monitoring/src/configuration.rs [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+
+use std::path::{Path, PathBuf};
+
+use clap::{App, Arg};
+
+pub struct DeployMonitoringEnvironment {
+    // logging level
+    pub log_level: slog::Level,
+
+    // slack bot token
```
Comment (bkontur) [inline apps/deploy_monitoring/src/configuration.rs:12]: @adonagy 
I would put this slack stuff to separate struct:
`struct SlackConfiguration` and also do it as optional here:
```
...
slack: Option<SlackConfiguration>
...
```
I want to run it locally without sending messages to any slack

File: apps/deploy_monitoring/src/configuration.rs [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+// Validates single required arg. If missing, exit whole process
+pub fn validate_required_arg(args: &clap::ArgMatches, arg_name: &str) {
+    if !args.is_present(arg_name) {
+        panic!("required \"{}\" arg is missing !!!", arg_name);
+    }
+}
+
+fn validate_required_args(args: &clap::ArgMatches) {
+    validate_required_arg(args, "image-monitor-interval");
+    // validate_required_arg(args, "resource-monitor-interval");
```
Comment (bkontur) [inline apps/deploy_monitoring/src/configuration.rs:188]: @adonagy 
if not needed please remove

File: apps/deploy_monitoring/src/deploy_with_compose.rs [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+use slog::{info, Logger};
+use tokio::time::{sleep, Duration};
+
+use crate::image::{DeployMonitoringContainer, Explorer, Sandbox, TezedgeDebugger};
+use crate::node::{OcamlNode, TezedgeNode, OCAML_PORT, TEZEDGE_PORT};
+
+pub const DEBUGGER_PORT: u16 = 17732;
+
+// TODO: use external docker-compose for now, should we manage the images/containers directly?
```
Comment (bkontur) [inline apps/deploy_monitoring/src/deploy_with_compose.rs:16]: @adonagy 
please create jira for leftover TODOs

File: apps/deploy_monitoring/src/deploy_with_compose.rs [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+pub async fn launch_stack(compose_file_path: &PathBuf, log: &Logger) {
```
Comment (bkontur) [inline apps/deploy_monitoring/src/deploy_with_compose.rs:17]: @adonagy 
Could you please clean this function a little bit, like:
```
log(X1 is starting)
start_with_compose
wait_for_start(...)
log(X1 is running)

log(X2 is starting)
start_with_compose
wait_for_start(...)
log(X2 is running)

log(X3 is starting)
start_with_compose
wait_for_start(...)
log(X3 is running)
```

File: apps/deploy_monitoring/src/deploy_with_compose.rs [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+    ))
+    .await
+    .is_err()
+    {
+        sleep(Duration::from_millis(1000)).await;
+    }
+    info!(log, "Ocaml node is running");
+}
+
+pub async fn launch_sandbox(compose_file_path: &PathBuf, log: &Logger) {
```
Comment (bkontur) [inline apps/deploy_monitoring/src/deploy_with_compose.rs:57]: @adonagy 
also please some cleaning as launch_stack

File: apps/deploy_monitoring/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+
+    let slack_server = slack::SlackServer::new(
+        env.slack_url,
+        env.slack_token,
+        env.slack_channel_name,
+        log.clone(),
+    );
+
+    let running = Arc::new(AtomicBool::new(true));
+    let mut thread_handles = Vec::new();
```
Comment (bkontur) [inline apps/deploy_monitoring/src/main.rs:43]: @adonagy 
I would maybe refactored it to separate fuctionsm which returns thread_handles, like:
```
let thread_handles = if env.is_sandbox {
    launch_sandbox(...)
} else {
    launch_node_stack(...)
};
```

File: apps/deploy_monitoring/src/monitors/alerts.rs [hunk 8]
```diff
@@ -0,0 +1,606 @@
...
+#[derive(Clone, Debug, Eq)]
+pub struct MonitorAlert {
+    level: AlertLevel,
+    kind: AlertKind,
+    timestamp: Option<i64>,
+    reported: bool,
+    value: u64,
+}
+
+// implement PartialEq and Hash traits manually to achieve that each kind of alert is only once in the HashSet
```
Comment (bkontur) [inline apps/deploy_monitoring/src/monitors/alerts.rs:42]: @adonagy 
I think, not necessery comment

File: apps/deploy_monitoring/src/monitors/alerts.rs [hunk 9]
```diff
@@ -0,0 +1,606 @@
...
+        let expected = MonitorAlert {
+            kind: AlertKind::Disk,
+            level: AlertLevel::Critical,
+            timestamp: None,
+            reported: false,
+            value: 300_000_000_000,
+        };
+
+        assert_eq!(disk_alert.level, expected.level);
+        drop(disk_alert);
```
Comment (bkontur) [inline apps/deploy_monitoring/src/monitors/alerts.rs:402]: @adonagy 
clippy error here, just remove drop

File: apps/deploy_monitoring/src/monitors/deploy.rs [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+                }
+                Err(e) => {
+                    bail!("Error collecting last: {:?}", e);
+                }
+            };
+
+        // "move" trough all the possible cursor_ids with a LIMIT step
+        while let Some(cursor) = cursor_id {
+            match reqwest::get(&format!(
+                "http://localhost:17732/v2/log?node_name=9732&limit={}&cursor_id={}",
```
Comment (bkontur) [inline apps/deploy_monitoring/src/monitors/deploy.rs:89]: @adonagy 
this should not be hard-coded here

File: apps/deploy_monitoring/src/monitors/deploy.rs [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+            log,
+            compose_file_path,
+            ..
+        } = self;
+
+        if self.is_node_container_running().await {
+            let node_updated = self.changed::<TezedgeNode>().await?;
+            let debugger_updated = self.changed::<TezedgeDebugger>().await?;
+            let explorer_updated = self.changed::<Explorer>().await?;
+            // TODO: here restart everything,
```
Comment (bkontur) [inline apps/deploy_monitoring/src/monitors/deploy.rs:134]: @adonagy 
add please leftover jira TODO here

File: apps/deploy_monitoring/src/monitors/info.rs [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+        let debugger_commit_hash = TezedgeDebugger::collect_commit_hash().await?; // same as OcamlDebugger
+        let explorer_commit_hash = Explorer::collect_commit_hash().await?;
+        let commit_hashes = CommitHashes::new(
+            ocaml_commit_hash,
+            tezedge_commit_hash,
+            debugger_commit_hash,
+            explorer_commit_hash,
+        );
+
+        // TODO fix, or remove, when removing slack info
```
Comment (bkontur) [inline apps/deploy_monitoring/src/monitors/info.rs:81]: @adonagy 
what about this?
please add at least leftover jira TODO

File: apps/deploy_monitoring/src/monitors/mod.rs [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+};
+use crate::monitors::deploy::DeployMonitor;
+use crate::monitors::resource::{ResourceMonitor, ResourceUtilizationStorage};
+use crate::slack::SlackServer;
+
+pub mod alerts;
+pub mod deploy;
+pub mod resource;
+
+// TODO: get this info from docker (shiplift needs to implement docker volume inspect)
```
Comment (bkontur) [inline apps/deploy_monitoring/src/monitors/mod.rs:27]: @adonagy 
leftover jira TODO please

File: apps/deploy_monitoring/src/monitors/mod.rs [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+            sleep(Duration::from_secs(interval)).await;
+        }
+    })
+}
+
+pub fn start_resource_monitoring(
+    interval: u64,
+    log: Logger,
+    running: Arc<AtomicBool>,
+    ocaml_resource_utilization: ResourceUtilizationStorage,
```
Comment (bkontur) [inline apps/deploy_monitoring/src/monitors/mod.rs:80]: @adonagy 
to be more flexible for future sandbox multinode support or single node run,
I would change this to: `resource_utilization: Vec<ResourceUtilizationStorage>`

File: apps/deploy_monitoring/src/monitors/resource.rs [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+}
+
+#[derive(Clone, Debug, Default, Serialize, Getters)]
+pub struct MemoryStats {
+    #[get = "pub(crate)"]
+    node: ProcessMemoryStats,
+
+    #[get = "pub(crate)"]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    protocol_runners: Option<Vec<ProcessMemoryStats>>,
```
Comment (bkontur) [inline apps/deploy_monitoring/src/monitors/resource.rs:45]: @adonagy 
remove protocol_runners and use validators for ocaml and tezedge

File: apps/deploy_monitoring/src/monitors/resource.rs [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+use crate::display_info::DiskData;
+use crate::node::{Node, TezedgeNode, OCAML_PORT, TEZEDGE_PORT};
+
+pub type ResourceUtilizationStorage = Arc<RwLock<VecDeque<ResourceUtilization>>>;
+
+/// The max capacity of the VecDeque holding the measurements
+pub const MEASUREMENTS_MAX_CAPACITY: usize = 40320;
+
+pub struct ResourceMonitor {
+    ocaml_resource_utilization: ResourceUtilizationStorage,
```
Comment (bkontur) [inline apps/deploy_monitoring/src/monitors/resource.rs:29]: @adonagy 
change here to `resource_utilization: Vec<ResourceUtilizationStorage>`

File: apps/deploy_monitoring/src/monitors/resource.rs [hunk 17]
```diff
@@ -0,0 +1,204 @@
...
+            tezedge_resource_utilization,
+            last_checked_head_level,
+            alerts,
+            log,
+            slack,
+            system: System::new_all(),
+        }
+    }
+
+    pub async fn take_measurement(&mut self) -> Result<(), failure::Error> {
```
Comment (bkontur) [inline apps/deploy_monitoring/src/monitors/resource.rs:97]: @adonagy 
I would not mix measurements,
if you refactor it to `resource_utilization: Vec<ResourceUtilizationStorage>`,
then you could iterate it and `take_measurement` for every entry

File: apps/deploy_monitoring/src/node.rs [hunk 18]
```diff
@@ -0,0 +1,192 @@
...
+pub const TEZEDGE_PORT: u16 = 18732;
+pub const OCAML_PORT: u16 = 18733;
+
+pub struct TezedgeNode;
+
+#[async_trait]
+impl Node for TezedgeNode {
+    fn collect_disk_data() -> Result<DiskData, failure::Error> {
+        let disk_data = TezedgeDiskData::new(
+            // dir::get_size(&format!("{}/{}", TEZEDGE_VOLUME_PATH, "debugger_db"))?,
```
Comment (bkontur) [inline apps/deploy_monitoring/src/node.rs:30]: @adonagy
???

File: apps/deploy_monitoring/src/node.rs [hunk 19]
```diff
@@ -0,0 +1,192 @@
...
+    }
+}
+
+pub struct OcamlNode;
+
+#[async_trait]
+impl Node for OcamlNode {
+    fn collect_disk_data() -> Result<DiskData, failure::Error> {
+        Ok(OcamlDiskData::new(
+            // dir::get_size(&format!("{}/{}", OCAML_VOLUME_PATH, "debugger_db"))?,
```
Comment (bkontur) [inline apps/deploy_monitoring/src/node.rs:85]: @adonagy 
???

File: apps/deploy_monitoring/src/rpc/handlers.rs [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+pub async fn get_measurements(
+    options: MeasurementOptions,
+    log: Logger,
+    measurements_storage: ResourceUtilizationStorage,
+) -> Result<impl warp::Reply, reject::Rejection> {
+    let storage = measurements_storage.read().unwrap();
+
+    info!(log, "Serving measurements");
```
Comment (bkontur) [inline apps/deploy_monitoring/src/rpc/handlers.rs:29]: @adonagy 
do we need this log?

File: apps/deploy_monitoring/src/rpc/mod.rs [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+
+use crate::monitors::resource::ResourceUtilizationStorage;
+
+pub mod filters;
+pub mod handlers;
+
+pub fn spawn_rpc_server(
+    rpc_port: u16,
+    log: Logger,
+    ocaml_resource_utilization: ResourceUtilizationStorage,
```
Comment (bkontur) [inline apps/deploy_monitoring/src/rpc/mod.rs:16]: @adonagy 
I would refactor it to:
```
resource_utilization: Vec<Arc<ResourceUtilizationStorage>>
```
to aviod lots of cloning

**Review by bkontur — CHANGES_REQUESTED** [review]
@adonagy 
please see PR comments

**Comment by adonagy** [comment]
@bkontur 
Thank you. I'll make the requested changes.

File: apps/deploy_monitoring/src/monitors/deploy.rs (hunk 11 above)
Comment (bkontur) [reply apps/deploy_monitoring/src/monitors/deploy.rs:136]: @adonagy Ado, what about this TODO?
remove it or add here TE-499 and add TODO here https://viablesystems.atlassian.net/browse/TE-499

File: apps/deploy_monitoring/src/monitors/alerts.rs [hunk 22]
```diff
@@ -0,0 +1,627 @@
...
+            kind: AlertKind::Disk,
+            level: AlertLevel::Info,
+            timestamp: None,
+            reported: false,
+            value: 40_000_000_001,
+        };
+
+        assert_eq!(disk_alert, &expected);
+        assert_eq!(disk_alert.level, expected.level);
+        drop(disk_alert);
```
Comment (bkontur) [inline apps/deploy_monitoring/src/monitors/alerts.rs:471]: @adonagy 
clippy error, remove this drop please

File: apps/deploy_monitoring/src/deploy_with_compose.rs [hunk 23]
```diff
@@ -0,0 +1,157 @@
...
+
+use crate::image::{DeployMonitoringContainer, Explorer, Sandbox, TezedgeDebugger};
+use crate::node::{OcamlNode, TezedgeNode, OCAML_PORT, TEZEDGE_PORT};
+
+pub const DEBUGGER_PORT: u16 = 17732;
+
+pub async fn launch_stack(compose_file_path: &PathBuf, log: &Logger) {
+    info!(log, "Tezedge explorer is starting");
+    start_with_compose(compose_file_path, Explorer::NAME, "explorer");
+    wait_for_start(&format!("http://localhost")).await;
```
Comment (bkontur) [inline apps/deploy_monitoring/src/deploy_with_compose.rs:19]: @adonagy 
clippy, just change `(&format!("http://localhost")` to `"http://localhost"`,
explorer does not have port?
I guess it take 80 as default, so I would prefer to use a constant here with 80

File: apps/deploy_monitoring/src/deploy_with_compose.rs [hunk 24]
```diff
@@ -0,0 +1,157 @@
...
+use std::path::PathBuf;
+use std::process::{Command, Output};
+
+use slog::{info, Logger};
+use tokio::time::{sleep, Duration};
+
+use crate::image::{DeployMonitoringContainer, Explorer, Sandbox, TezedgeDebugger};
+use crate::node::{OcamlNode, TezedgeNode, OCAML_PORT, TEZEDGE_PORT};
+
+pub const DEBUGGER_PORT: u16 = 17732;
```
Comment (bkontur) [inline apps/deploy_monitoring/src/deploy_with_compose.rs:14]: @adonagy 
I would put all (at least port) constants to one file/mod (it shoudl be better as configuration, but constants is enough):
OCAML_PORT, TEZEDGE_PORT, DEBUGGER_PORT

File: apps/deploy_monitoring/src/monitors/deploy.rs [hunk 25]
```diff
@@ -0,0 +1,250 @@
...
+};
+use crate::image::{
+    local_hash, remote_hash, DeployMonitoringContainer, Explorer, Sandbox, TezedgeDebugger,
+};
+
+use crate::monitors::TEZEDGE_VOLUME_PATH;
+use crate::node::TezedgeNode;
+use crate::slack::SlackServer;
+
+const TEZEDGE_NODE_P2P_PORT: u16 = 9732;
```
Comment (bkontur) [inline apps/deploy_monitoring/src/monitors/deploy.rs:28]: @adonagy 
also port constant, I would put them together in one place, maybe `mod constants`

File: apps/deploy_monitoring/Cargo.toml [hunk 26]
```diff
@@ -0,0 +1,33 @@
...
+serde_json = "1.0"
+shiplift = { git = "https://github.com/simplestaking/shiplift.git", branch = "master" }
+slog = { version = "2.5", features = ["nested-values"] }
+slog-async = "2.5"
+slog-term = "2.6"
+sysinfo = "0.16"
+tokio = { version = "1.2", features = ["full"] }
+wait-timeout = "0.2"
+warp = "0.3"
+shell = { path = "../../shell" }
```
Comment (bkontur) [inline apps/deploy_monitoring/Cargo.toml:32]: @adonagy 
please add here `TODO: TE-499 remove shell dependency, and move stats/memory somewhere`

File: apps/deploy_monitoring/src/monitors/resource.rs [hunk 27]
```diff
@@ -0,0 +1,216 @@
...
+#[derive(Clone, Debug, Serialize, Getters, Merge, Default)]
+pub struct MemoryStats {
+    #[get = "pub(crate)"]
+    #[merge(strategy = merge::ord::max)]
+    node: ProcessMemoryStatsMaxMerge,
+
+    #[get = "pub(crate)"]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[merge(strategy = merge::ord::max)]
+    protocol_runners: Option<ProcessMemoryStatsMaxMerge>,
```
Comment (bkontur) [inline apps/deploy_monitoring/src/monitors/resource.rs:49]: @adonagy 
please add here `TODO: TE-499 remove protocol_runners and use validators for ocaml and tezedge type`

**Review by bkontur — CHANGES_REQUESTED** [review]
@adonagy 
ok, just few cosmetics left, check them please,
and also check rustfmt complains,
and please squash a little bit commits

File: apps/deploy_monitoring/src/deploy_with_compose.rs (hunk 23 above)
Comment (adonagy) [reply apps/deploy_monitoring/src/deploy_with_compose.rs:19]: Exactly, the explorer uses 80, I'll add it as a constant

File: apps/deploy_monitoring/src/constants/mod.rs [hunk 28]
```diff
@@ -0,0 +1,18 @@
+// Port constants
```
Comment (bkontur) [inline apps/deploy_monitoring/src/constants/mod.rs:1]: @adonagy 
missing 
```
// Copyright (c) SimpleStaking and Tezedge Contributors
// SPDX-License-Identifier: MIT
```

and I would just create constants.rs insetad of subdir constants/mod.rs

**Pull request merged**